File: Learning/src/Questions/DP/CoinChange.java
package Questions.DP;

import java.util.Arrays;

public class CoinChange {
    // Problem:- We are given an array Arr with N distinct coins and a target. We
    // have an infinite supply of each coin denomination. We need to find the number
    // of ways we sum up the coin values to give us the target.

    // Recursive function to count the ways to make change
    static long countWaysToMakeChangeUtil(int[] arr, int ind, int T, long[][] dp) {
        // Base case: If the current index is 0
        if (ind == 0) {
            // If T is divisible by the first element of the array, return 1, else return 0
            if (T % arr[0] == 0)
                return 1;
            else
                return 0;
        }

        // If the result for this subproblem has already been calculated, return it
        if (dp[ind][T] != -1)
            return dp[ind][T];

        // Calculate the number of ways without taking the current element
        long notTaken = countWaysToMakeChangeUtil(arr, ind - 1, T, dp);

        // Initialize the number of ways taking the current element as 0
        long taken = 0;

        // If the current element is less than or equal to T, calculate 'taken'
        if (arr[ind] <= T)
            taken = countWaysToMakeChangeUtil(arr, ind, T - arr[ind], dp);

        // Store the result in the dp array and return it
        return dp[ind][T] = notTaken + taken;
    }

    // Function to count the ways to make change
    static long countWaysToMakeChangeRecursive(int[] arr, int n, int T) {
        // Create a 2D array to store results of subproblems
        long dp[][] = new long[n][T + 1];

        // Initialize the dp array with -1 to indicate that subproblems are not solved
        // yet
        for (long row[] : dp)
            Arrays.fill(row, -1);

        // Call the countWaysToMakeChangeUtil function to calculate the number of ways
        return countWaysToMakeChangeUtil(arr, n - 1, T, dp);
    }
    // Time Complexity: O(N*T)

    // Reason: There are N*W states therefore at max ‘N*T’ new problems will be
    // solved.

    // Space Complexity: O(N*T) + O(N)

    // Reason: We are using a recursion stack space(O(N)) and a 2D array ( O(N*T)).

    // Function to count the ways to make change
    static long countWaysToMakeChangeTabulation(int[] arr, int n, int T) {
        // Each cell dp[i][j] stores the number of ways to make change for the target
        // 'j' using the first 'i' coins
        long dp[][] = new long[n][T + 1];

        // Initialize the first column of the dp array
        // If the target is divisible by the first element of the array, set the value
        // to
        // 1, else set it to 0
        for (int i = 0; i <= T; i++) {
            if (i % arr[0] == 0)
                dp[0][i] = 1;
            // Else condition is automatically fulfilled, as dp array is initialized to zero
        }

        // Fill the dp array using dynamic programming
        for (int ind = 1; ind < n; ind++) {
            for (int target = 0; target <= T; target++) {
                long notTaken = dp[ind - 1][target];

                long taken = 0;
                if (arr[ind] <= target)
                    taken = dp[ind][target - arr[ind]];

                dp[ind][target] = notTaken + taken;
            }
        }

        return dp[n - 1][T];
    }
    // Time Complexity: O(N*T)

    // Reason: There are two nested loops

    // Space Complexity: O(N*T)

    // Reason: We are using an external array of size ‘N*T’. Stack Space is
    // eliminated.

    // Function to count the ways to make change
    static long countWaysToMakeChangeOptimized(int[] arr, int n, int T) {
        // Create an array to store results of subproblems for the previous element
        long[] prev = new long[T + 1];

        // Initialize base condition for the first element of the array
        for (int i = 0; i <= T; i++) {
            if (i % arr[0] == 0)
                prev[i] = 1;
            // Else condition is automatically fulfilled, as prev array is initialized to
            // zero
        }

        // Fill the prev array using dynamic programming
        for (int ind = 1; ind < n; ind++) {
            // Create an array to store results of subproblems for the current element
            long[] cur = new long[T + 1];
            for (int target = 0; target <= T; target++) {
                long notTaken = prev[target];

                long taken = 0;
                if (arr[ind] <= target)
                    taken = cur[target - arr[ind]];

                cur[target] = notTaken + taken;
            }
            prev = cur;
        }

        return prev[T];
    }
    // Time Complexity: O(N*T)

    // Reason: There are two nested loops.

    // Space Complexity: O(T)

    // Reason: We are using an external array of size ‘T+1’ to store two rows only.
    public static void main(String[] args) {
        int arr[] = { 1, 2, 3 };
        int N = arr.length;
        int T = 4;

        System.out.println(countWaysToMakeChangeRecursive(arr, N, T));
        System.out.println(countWaysToMakeChangeTabulation(arr, N, T));
        System.out.println(countWaysToMakeChangeOptimized(arr, N, T));
    }

}

File: DSA/Python/Questions/DP/coin_change.py
"""
# Question: Coin Change
# Link: https://leetcode.com/problems/coin-change/

# Problem Statement:
# Given an array of distinct coins and a target sum, find the number of ways
# to make change for the target using the coins. You have an infinite supply
# of each coin denomination.

# Example:
# Input: coins = [1,2,3], target = 4
# Output: 4
# Explanation: {1,1,1,1}, {1,1,2}, {2,2}, {1,3}

# Key Insights:
# - Each coin can be used multiple times (infinite supply)
# - Order of coins in combination doesn't matter
# - Need to handle overlapping subproblems efficiently
"""

from typing import List

# 1. Recursive Solution with Memoization
"""
Algorithm:
1. For each coin, we have two choices at each step:
   - Take the current coin if target >= coin value
   - Skip the current coin and move to next
2. Use memoization to avoid recalculating same states

Time Complexity: O(N*T) where N is number of coins and T is target
Space Complexity: O(N*T) for dp array + O(N) for recursion stack
"""


def count_ways_recursive(coins: List[int], target: int) -> int:
    # Initialize dp array for memoization
    n = len(coins)
    dp = [[-1] * (target + 1) for _ in range(n)]

    def solve(ind: int, target: int) -> int:
        # Base case: found valid combination
        if target == 0:
            return 1

        # Base case: processing first coin
        if ind == 0:
            # Check if target is divisible by first coin
            return 1 if target % coins[0] == 0 else 0

        # Return memoized result if available
        if dp[ind][target] != -1:
            return dp[ind][target]

        # Calculate ways without taking current coin
        not_take = solve(ind - 1, target)

        # Calculate ways by taking current coin if possible
        take = 0
        if coins[ind] <= target:
            take = solve(ind, target - coins[ind])

        # Store result in dp array and return
        dp[ind][target] = take + not_take
        return dp[ind][target]

    return solve(n - 1, target)


# 2. Tabulation Solution
"""
Algorithm:
1. Build solution bottom-up using 2D dp array
2. For each coin and target value:
   - Calculate ways without current coin
   - Add ways including current coin if possible

Time Complexity: O(N*T)
Space Complexity: O(N*T)
"""


def count_ways_tabulation(coins: List[int], target: int) -> int:
    n = len(coins)
    # Create and initialize dp array
    dp = [[0] * (target + 1) for _ in range(n)]

    # Initialize first row (base case for first coin)
    for t in range(target + 1):
        dp[0][t] = 1 if t % coins[0] == 0 else 0

    # Fill dp table for remaining coins
    for ind in range(1, n):
        for t in range(target + 1):
            # Ways without current coin
            not_take = dp[ind - 1][t]

            # Ways including current coin
            take = 0
            if coins[ind] <= t:
                take = dp[ind][t - coins[ind]]

            # Total ways = ways with + ways without
            dp[ind][t] = take + not_take

    return dp[n - 1][target]


# 3. Space Optimized Solution
"""
Algorithm:
1. Use only previous row to calculate current row
2. Optimize space by using 1D arrays instead of 2D

Time Complexity: O(N*T)
Space Complexity: O(T)
"""


def count_ways_optimized(coins: List[int], target: int) -> int:
    n = len(coins)
    # Initialize previous array with base case
    prev = [1 if t % coins[0] == 0 else 0 for t in range(target + 1)]

    # Process each coin
    for ind in range(1, n):
        # Current state array
        curr = [0] * (target + 1)

        # Calculate for each target
        for t in range(target + 1):
            # Ways without current coin
            not_take = prev[t]

            # Ways with current coin
            take = 0
            if coins[ind] <= t:
                take = curr[t - coins[ind]]

            # Update current state
            curr[t] = take + not_take

        # Update previous for next iteration
        prev = curr[:]

    return prev[target]


def main():
    # Test cases covering various scenarios
    test_cases = [
        ([1, 2, 3], 4),  # Multiple solutions
        ([2, 5, 10], 15),  # Larger target
        ([1], 5),  # Single coin
        ([1, 2, 5], 0),  # Zero target
        ([186, 419, 83, 408], 6249),  # Large numbers
    ]

    # Run all solutions for each test case
    for i, (coins, target) in enumerate(test_cases):
        print(f"\nTest Case {i + 1}: coins = {coins}, target = {target}")
        print(f"Recursive Solution: {count_ways_recursive(coins, target)}")
        print(f"Tabulation Solution: {count_ways_tabulation(coins, target)}")
        print(f"Optimized Solution: {count_ways_optimized(coins, target)}")


if __name__ == "__main__":
    main()
